tests/Unit/Parallel/CMakeLists.txt

```diff
@@ -110,6 +110,37 @@ function(add_algorithm_test_with_input_file TEST_NAME INPUT_FILE_DIR)
       ENVIRONMENT "ASAN_OPTIONS=detect_leaks=0")
 endfunction()
 
+# Note that we have a 'balance' here that happens 2e4 of times per second just
+# to be completely certain that a migration will happen during the execution so
+# we can verify the pup
+function(add_algorithm_test_with_balancing TEST_NAME EXECUTABLE_NAME REGEX_TO_MATCH)
+  add_test(
+    NAME "\"Integration.Parallel.${TEST_NAME}\""
+    COMMAND
+    ${SHELL_EXECUTABLE}
+    -c "${CMAKE_BINARY_DIR}/bin/Test_${EXECUTABLE_NAME} +p2 +balancer \
+     RandCentLB +LBPeriod 5.0e-5 +LBDebug 3 2>&1"
```

Comment:
I have a couple concerns with this test.
- This won't work with C++ modules (whenever we get to switch to them) because modules requires that you don't switch definitions locally based on macros. We can do this as a temporary hold over until more infrastructure gets in, but basically this can't really stay in the code very long otherwise it'll get forgotten and make somebody's life really terrible when they try to switch the code to modules.
- Why not use the `RotateLB`? The description for that is: `This strategy moves objects to the next available PE every time it is called. It is useful for debugging PUP routines and other migration related bugs.` and that's exactly what we want to do here: debug/test PUP routines. We could stick something in the DataBox and Inbox, then make sure things get migrated correctly. I think we can test the phase, databox, algorithm step, and inboxes (maybe the array_index too?) via some action calls. I'm not sure how to test the global cache serialization without going to multinode, but we can at least verify that we can retrieve a correct value out of the cache both explicitly and through the DataBox. 
I don't know how much of the suggested test can be done with what's already in the code, so this is mainly a discussion point to see what we can do now and what we ultimately I think should do for testing this.

Reply:
Gotcha, that's a good point regarding modules. The main problem I'm trying to solve with that trick is 'how do we trust that charm has really called the pup function'. I think the only way we can truly verify that is to extract some debugging information from the pup function or from extra debug functions in the AlgorithmImpl, or by manually running the serialization ourselves.

Regarding `RotateLB` - I tried that first, but I couldn't manage to get it to consistently migrate on two cores. I can't tell if that's a charm problem or a failure in my understanding.

Here's an alternative idea that shouldn't be too painful to code up -- in the `Test_AlgorithmParallel.cpp` test, add an action for the array that when the number of processors is exactly one (so we can be sure that other elements on the same core won't be simultaneously running), manually serializes/deserializes every other element of the array in circular permutation, then continues with the rest of the algorithm test. Does that sound robust enough to not need to be rewritten in the near future?

The only other note there is that it requires that at least one case of `Test_AlgorithmParallel.cpp` to be run on a single core, though I think that's how it works in the current cmake now anyways?

Reply:
Uh oh, this is spooky... Does `RotateLB` work on more than two cores? I think that actually needs to be understood because if we're seeing "weird load-balancing behavior" we should at least be able to do a rotate of the elements across cores correctly independent of the number of cores. If there is an issue on a specific number of cores we should be sure it's not how we use LBs because if we encounter other issues than we can't be sure why.

I think doing the explicit serialize/deserialize in `Test_AlgorithmParallel.cpp` should work. You're saying get the local object, serialize/deserialize and then do what with it? This seems like it could work, but I think we should run it through the sanitizers to make sure Charm doesn't make some assumption that we violate